src/ctap/apdu.rs

```diff
@@ -120,20 +124,23 @@ impl TryFrom<&[u8]> for APDU {
         let (header, payload) = frame.split_at(APDU_HEADER_LEN);
 
         let mut apdu = APDU {
-            header: header.into(),
+            header: array_ref!(header, 0, 4).into(),
             lc: 0x00,
             data: Vec::new(),
             le: 0x00,
             case_type: ApduType::default(),
         };
 
-        // case 1
         if payload.is_empty() {
+            // Lc is zero-bytes in length
             apdu.case_type = ApduType::Instruction;
         } else {
```

Comment:
General question, since this is a pretty big if-else-construction. The structure is
- `case_type: ApduType::default()`
- lots of `if-else`
- check `case_type`, return error if `Unknown`
- return `apdu`

Would it be more readable if we early return in all known cases, and then return the error as a default in the end? That might mean to not have a `mut apdu`, but actually construct `APDU`s for returning in each case. WDYT?

Reply:
That's a pretty great idea! Mutation is evil :) , addressed in 1d8c103d9b8180b93994e35ea9365c780e15803b

Reply:
Nice! I'd personally remove the else, since the if block before always returns.

Reply:
Also, `ApduType` doesn't seem to need `default` and `Unknown` anymore?

Reply:
Done

Reply:
Sorry, I had two comments in one thread. What do you think about removing indentation levels after you return?

Reply:
done, addressed in 4bfce88